### ISSUE TITLE
Provide UIElements with the ability to set the mouse cursor

### ISSUE THREAD

**Issue by abaga129** [AuburnSounds/Dplug#424]
It would be great to have a way for UIElements to set the cursor so that they can convey more purpose to the user.

A couple of examples where this could be useful:
- in `onMouseEnter` of a slider, the mouse would change to a pointer, and in `onMouseLeave` it would return to the default.
![drag cursor](https://docs.microsoft.com/en-us/windows/win32/uxguide/images/inter-mouse-image10.png)
- In M4 Multiband Compressor, I would like the drag-able crossovers to show a `resize` cursor like the one below
![Resize Cursor](https://docs.microsoft.com/en-us/windows/win32/uxguide/images/inter-mouse-image12.png)
- For the browser links we could also indicate that the area is actually a clickable link with something like this
![link select cursor](https://docs.microsoft.com/en-us/windows/win32/uxguide/images/inter-mouse-image6.png)

These are just examples from the [win32 docs](https://docs.microsoft.com/en-us/windows/win32/uxguide/inter-mouse), but each of the windowing solutions should expose some way to set these.

I plan to work on this at some point this year, but if someone else would like to work on this I'm putting a $80 bounty on it.  See the [Bug Bounty Program](https://github.com/AuburnSounds/Dplug/wiki/Bug-Bounty-Program) for details on this.

**Comment by p0nce** [comment]
Good idea. Some plugins also hide the mouse in some situations (dragging a knob).

**Comment by abaga129** [comment]
I began working on this a little while back but it still needs some work.  It gets tricky when you throw dragging into the mix.  I'm still more than happy to pay for this to be implemented.  Otherwise I'll get to it eventually. 😃

**Comment by p0nce** [comment]
I think changing the cursor would usually go into `onBeginDrag` and `onStopDrag`, I have not though about how to support onMouseEnter / onMouseLeave.

In any case, a custor "stack" (or priority queue?) will probably need to be implemented in the `UIContext`, so tht ending a drag can restore the previous one. 
It gets tricky since onStopDrag can occur at odd times, after the mouse has leaved the area.

**Comment by abaga129** [comment]
Yeah that's pretty much the issue I was running into.  I have a `Cursor` property on the UIContext, but I hadnt thought of using a stack.  My issue was that if Element X  set the cursor to A in `onBeginDrag`, and then stopped dragging while the cursor was on top of Element Y.  Element Y could have logic to set the cursor to B `onMouseEnter`, but the cursor would go to whatever Element X said it should be in `onStopDrag`

**Comment by p0nce** [comment]
Without thinking too much about it, it seems we need a "priority queue".
Setting the cursor when dragging would set with a higher priority than setting it because of mouse over.

When the former cursor needs to be popped, what is popped is the state with 
Since the "pqueue" will be small, implementing it with a Vec should be enough I guess; the popCursor operation would find the item with highest priority.

Only one UIElement can be dragged or mouseOver at once, so the maximum item count of this thing could be 2 (or 3 with the "default state", always in the structure).

**Comment by abaga129** [comment]
I'm coming back to this one after a fairly long hiatus. I have most of the code implemented to handle this already. I'm going to work on writing a priority queue adapter using `Vec` like you suggested.  That should definitely simplify the problem.

**Comment by p0nce** [comment]
Er I was confused in my message above. 

Only one drag can be done at once, so I don't think you need a stack since UIContext points to the current "dragged" and "mouseOver" widgets. So if a drag is ending, you can still retrieve the active widget.

I thought about it a bit, and it seems to be a lot of widgets fall into the cases where there is 
- a changing cursor when mouse is over  (`isMouseOver`)
- and/or a changing cursor when mouse is dragging the widget (`isDragged`)

So I was thinking of 2 new UIElements members: one describing the cursor when dragged, and one describing the cursor when mouseOver. That way UICOntext can retrieve the cursor that was "mouseOver" after a drag is finished.

For example:


```d
class UIElement
{
   // blah
    MyCursorEnum _cursorWhenDragged;
    MyCursorEnum _cursorWhenMouseOver;
}
enum MyCursorEnum
{
    inherit,
    normal,
    hidden,
    rightArrow,
    hand,
}
```
something like that.


One would still need manual handling for more complex widgets with multiple things to drag though (eg: there are such widgets in Renegate), but that can come later.

**Comment by abaga129** [comment]
I think we could make it so that the MouseCursors are retrieved via methods.  By default on UIElement these could just return some private property, but they could easily be overriden by derived classes to have more complex behavior.

**Comment by p0nce** [comment]
Sounds great!

**Comment by p0nce** [comment]
This is now disabled since Dplug v9.2.16 unless you define the version identifier `futureMouseCursor`.
To do this you just have to add

```json
"versions": ["futureMouseCursor"],
```

in your `dub.json`
